hoomd/mpcd/RandomSubsetPicker.h

```diff
@@ -0,0 +1,292 @@
+        GPUArray<unsigned int> m_picks_idx;
+    };
+
+/*
+ * \param picks       //!< indexes that are picked in the *original* flags array
+ * \param N_pick      //!< number of particles picked
+ * \param flags       //!< flags indicating which particles to pick from [ 0 1 0 1 1 ...] 
+ * \param N_try_pick  //!< target number of particles to pick (if total number of 1's in flags will be less than N_try_pick, it will pick all the 1's, in that case N_pick would be less then N_try_pick
+ * \param seed        //!< Seed to random number Generator
+ * \param timestep    //!< timestep at which this is called
+ * \param N           //!< total number of particles 
+ */
+template<typename T>
+void RandomSubsetPicker::operator()(GPUArray<unsigned int>& picks,
+                                    unsigned int& N_pick,
+                                    const GPUArray<T>& flags,
+                                    unsigned int N_try_pick,
+                                    unsigned int timestep,
+                                    unsigned int N)
+    {
+    /*
+     * It first calculates total number of 1's in flags(N_total) and store the indices of those 1's in m_flags_idx.
+     * then it calculates total number of 1's on all the ranks(N_total_all_ranks).
+     * If N_total_all_ranks < N_try_pick it picks all the particles 
+     * else it does more complicated picking logic
+     *
+     *
+     * In the end it stores the indexes of picked particles(indices of picked particles according to the *original* flags array.) in picks and 
+     * number of particles picked to N_pick indexes that were picked in the *original* flags array.
+     */
+
+    /*
+     * First calculating How many marked particles are there in flags(1 = marked) and storing their indexes in m_flags_idx
+     * This is done by CalculateTotalNumbersToPickFrom()
+     */
+    unsigned int N_total = CalculateTotalNumbersToPickFrom<T>(flags, N);
+    unsigned int N_total_all_ranks = N_total;
+    unsigned int N_before = 0;
+
+    // reduce / scan the number of particles that are marked on all ranks
+    #ifdef ENABLE_MPI
+    if (m_exec_conf->getNRanks() > 1)
+        {
+        MPI_Allreduce(MPI_IN_PLACE, &N_total_all_ranks, 1, MPI_UNSIGNED, MPI_SUM, m_exec_conf->getMPICommunicator());
+        MPI_Exscan(&N_total, &N_before, 1, MPI_UNSIGNED, MPI_SUM, m_exec_conf->getMPICommunicator());
+        }
+    #endif // ENABLE_MPI
+
+    // If N_try_pick >= N_total_all_ranks - pick all particles
+    if (N_total_all_ranks <= N_try_pick)
+        {
+        m_picks_idx.resize(N_total);
+        ArrayHandle<unsigned int> h_picks_idx(m_picks_idx, access_location::host, access_mode::overwrite);
+        std::iota(h_picks_idx.data, h_picks_idx.data + N_total, 0);
+        m_Npick = N_total;
+        }
+    else
+        {
+        // do the pick logic
+        makeAllPicks(timestep, N_try_pick, N_total_all_ranks);
```

Comment:
I think it would make sense for this and the code that comes after it (lines 170-204) to all be in one function, since they are very closely related to each other.